### ISSUE TITLE
Logstash 2.4.0 unable to update major versions of plugins

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#5858]
Logstash 2.4.0 has been released with a backward compatible layer that allows plugins from the future logstash 5.0 to be installable.

However, it is not possible to upgrade a plugin beyond the installed major version of a plugin:

```
logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-filter-date
logstash-filter-date (2.1.6)
logstash-2.4.0$ bin/logstash-plugin update logstash-filter-date
No plugin updated
```

Even though [version 3.0.3](https://rubygems.org/gems/logstash-filter-date/versions/3.0.3) exists.
Currently the only workaround is to manually edit the Gemfile and remove the version constraint from that plugin.

```ruby
# from
gem "logstash-filter-date", "<3.0.0"
# to
gem "logstash-filter-date"
```

then it works:

````
logstash-2.4.0$ bin/logstash-plugin update logstash-filter-date 
Updated logstash-filter-date 2.1.6 to 3.0.3
```

Related issues: https://github.com/elastic/logstash/issues/5580 https://github.com/elastic/logstash/issues/5575

**Comment by jsvd** [comment]
A pontential fix for this is to modify the Gemfile to have the version constraints during package creating, but remove them before inserting the Gemfile into the package

**Comment by wiibaa** [comment]
https://github.com/logstash-plugins/logstash-output-kafka/issues/91 made me think that the behaviour is not the same for all plugins. It seems that plugin that are runtime-deps of others plugins behaves a little differently. Here are my findings


**TL;DR** With 2.4.0
* `bin/logstash-plugin update` seems broken and won't update **any plugin** to latest major/minor
* `bin/logstash-plugin install --version=X.X.X` works with latest major version for **all plugins expects 8 plugins that are runtime-deps of others** (filter-date,filter-grok,input-tcp,input-rabbitmq,input-generator,output-file,filter-json,output-elasticsearch)

### Example 1: "standard" plugin => logstash-output-kafka

Update => Do nothing
Install --version => **OK**
````
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-output-kafka
logstash-output-kafka (2.0.5)
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin update logstash-output-kafka
No plugin updated
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-output-kafka
logstash-output-kafka (2.0.5)
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin install --version=5.0.4 logstash-output-kafka
Validating logstash-output-kafka-5.0.4
Installing logstash-output-kafka
Installation successful
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-output-kafka
logstash-output-kafka (5.0.4)

````


### Example 2: "depended-on" plugin => logstash-input-generator

Update => Do nothing
Install --version => **KO**
````
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-input-generator
logstash-input-generator (2.0.4)
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin update logstash-input-generator
No plugin updated
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin install --version=3.02 logstash-input-generator
Validating logstash-input-generator-3.02
Plugin logstash-input-generator version 3.02 does not exist
ERROR: Installation aborted, verification failed for logstash-input-generator 3.02
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin install --version=3.0.2 logstash-input-generator
Validating logstash-input-generator-3.0.2
Installing logstash-input-generator
Plugin not found, aborting
ERROR: Installation Aborted, message: You have requested:
  logstash-input-generator = 3.0.2

The bundle currently has logstash-input-generator locked at 2.0.4.
Try running `bundle update logstash-input-generator`
````

### Example 3: minor version update of "standard" plugin => logstash-output-kafka

Downgrade then call update => No update
````
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin install --version=2.0.3 logstash-output-kafka
Validating logstash-output-kafka-2.0.3
Installing logstash-output-kafka
Installation successful
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin update logstash-output-kafka
No plugin updated
philippe@debian:~/dev/wiibaa/logstash/logstash-2.4.0$ bin/logstash-plugin list --verbose logstash-output-kafka
logstash-output-kafka (2.0.3)
````

**Comment by purbon** [comment]
Hi,
  catching up after some time off, here is one proposal to prevent regressions and test this behaviour. When we initially merged the [qa framework](https://github.com/elastic/logstash/tree/master/qa) it was desided to only have it for master and 5.0, what about having the same thing also for 2.x (aka 2.4 branch)?. 

Then we could add a test for this behaviour and prevent this type of problems. To do that we could update acoordingly the dummy gems available [inside](https://github.com/elastic/logstash/tree/master/qa/support/logstash-filter-qatest) the very same framework.

What do you think?
